### PR TITLE
#72 Component edit: add intellisense in cxml files - only state id

### PR DESCRIPTION
**Pull request by nicolaserny**
This PR is focused on state id intellisense.
The readme will be updated when the intellisense (on state properties) is over
#72

### CODE REVIEW

File: extension/src/completion/IdCompletionProvider.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        this.tagName = tagName;
+    }
+
+    public createSuggestion(document: vscode.TextDocument): vscode.CompletionItem[] {
+        const newId = this.getNewId(document);
+        return [new vscode.CompletionItem(newId.toString())];
+    }
+
+    private getNewId(document: vscode.TextDocument): number {
+        const regexToGetIds = new RegExp(`<${this.tagName}[\\s\\S]*? Id="(.*?)" [\\s\\S]*?>`, "g");
```
Comment (maurelio1234) [inline extension/src/completion/IdCompletionProvider.ts]: does it work if we change order of parameters?

File: extension/src/completion/componentCompletionItemProvider.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    ];
+
+    public provideCompletionItems(
+        document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken):
+        Thenable<vscode.CompletionItem[]> {
+        return new Promise<vscode.CompletionItem[]>((resolve, reject) => {
+            const documentTextRange = new vscode.Range(new vscode.Position(position.line, 0), position);
+            const textBeforeCursor = document.getText(documentTextRange);
+            const tagName = this.getTagName(textBeforeCursor);
+            const attribute = this.getAttributegName(textBeforeCursor);
```
Comment (maurelio1234) [inline extension/src/completion/componentCompletionItemProvider.ts]: getAttributeName

File: extension/src/completion/componentCompletionItemProvider.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    private completionProviders = [
+        { tag: "StateData", attribute: "Id", provider: new IdCompletionProvider("StateData") }
+    ];
+
+    public provideCompletionItems(
+        document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken):
+        Thenable<vscode.CompletionItem[]> {
+        return new Promise<vscode.CompletionItem[]>((resolve, reject) => {
+            const documentTextRange = new vscode.Range(new vscode.Position(position.line, 0), position);
+            const textBeforeCursor = document.getText(documentTextRange);
```
Comment (maurelio1234) [inline extension/src/completion/componentCompletionItemProvider.ts]: so it only works if tag and attributes are on the same line, right?

File: extension/src/completion/IdCompletionProvider.ts (hunk 1 above)
Comment (nicolaserny) [reply extension/src/completion/IdCompletionProvider.ts]: Yes :)

Comment (nicolaserny) [reply extension/src/completion/IdCompletionProvider.ts]: We have to more tests (I am not a regex expert).
But currently it is working pretty well

File: extension/src/completion/componentCompletionItemProvider.ts (hunk 3 above)
Comment (nicolaserny) [reply extension/src/completion/componentCompletionItemProvider.ts]: Yes I know this limitation. I have an idea for this fix but will do it later on

File: extension/src/completion/componentCompletionItemProvider.ts (hunk 2 above)
Comment (nicolaserny) [reply extension/src/completion/componentCompletionItemProvider.ts]: done :)

**Pull request merged**